src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,205 @@
+    M<:OptimLogDensity
+} <: StatsBase.StatisticalModel
+    "A vector with the resulting point estimates."
+    values :: V
+    "The stored Optim.jl results."
+    optim_result :: O
+    "The final log likelihood or log joint, depending on whether `MAP` or `MLE` was run."
+    lp :: Float64
+    "The evaluation function used to calculate the output."
+    f :: M
+end
+
+#############################
+# Various StatsBase methods #
+#############################
+
+function Base.show(io::IO, m::ModeResult)
+    println(io, Printf.@sprintf("ModeResult with minimized lp of %.2f", m.lp), "\n")
+    show(io, m.values)
+end
+
+function StatsBase.coeftable(m::ModeResult)
+    # Get columns for coeftable.
+    terms = StatsBase.coefnames(m)
+    estimates = m.values.array[:,1]
+    stderrors = StatsBase.stderror(m)
+    tstats = estimates ./ stderrors
+
+    StatsBase.CoefTable([estimates, stderrors, tstats], ["estimate", "stderror", "tstat"], terms)
+end
+
+function StatsBase.informationmatrix(m::ModeResult; hessian_function=ForwardDiff.hessian, kwargs...)
+    # Calculate Hessian and information matrix.
+    varnames = StatsBase.coefnames(m)
+    info = hessian_function(x -> m.f(x, unlinked=true), m.values.array[:, 1])
+    info = inv(info)
+    return NamedArrays.NamedArray(info, (varnames, varnames))
+end
+
+StatsBase.coef(m::ModeResult) = m.values
+StatsBase.coefnames(m::ModeResult) = names(m.values)[1]
+StatsBase.params(m::ModeResult) = StatsBase.coefnames(m)
+StatsBase.vcov(m::ModeResult) = StatsBase.informationmatrix(m)
+StatsBase.loglikelihood(m::ModeResult) = m.lp
+
+####################
+# Optim.jl methods #
+####################
+
+"""
+    Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+
+Compute a maximum likelihood estimate of the `model`.
+
+Example:
+
+```julia
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+mle = optimize(model, MLE())
+
+# Use a different optimizer
+mle = optimize(model, MLE(), NelderMeade())
+```
+"""
+function Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+    return optimize(model, OptimLogDensity(model, DynamicPPL.LikelihoodContext()), args...; kwargs...)
+end
+
+"""
+    Optim.optimize(model::Model, ::MAP, args...; kwargs...)
+
+Compute a maximum a posterior estimate of the `model`.
+
+Example:
+
+```julia
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+map_est = optimize(model, MAP())
+
+# Use a different optimizer
+map_est = optimize(model, MAP(), NelderMeade())
+```
+"""
+function Optim.optimize(model::Model, ::MAP, args...; kwargs...)
+    return optimize(model, OptimLogDensity(model, DynamicPPL.DefaultContext()), args...; kwargs...)
+end
+
+"""
+    Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.BFGS(), args...; kwargs...)
+
+Estimate a mode, i.e., compute a MLE or MAP estimate.
+"""
+function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.BFGS(), args...; kwargs...)
```

Comment:
The memory requirement of BFGS scales quadratic in the number of parameter you optimise. I would suggest to use a more memory friendly option as default.
```suggestion
function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim. LBFGS(), args...; kwargs...)
```
or 
```suggestion
function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.GradientDescent(), args...; kwargs...)
```